src/lazy/encoder/binary/v1_1/value_writer.rs

```diff
@@ -231,8 +232,253 @@ impl<'value, 'top> BinaryValueWriter_1_1<'value, 'top> {
+            // nanoseconds field and write them out as a final byte. Because `subseconds` represents a number of
+            // nanoseconds between 0 and 999,999,999, its most significant 2 bits are guaranteed to be `00`. We can
+            // isolate the most significant 6 bits of the magnitude by isolating the most significant 8 bits of
+            // `subseconds`, which is a u32.
+            let high_six = (subseconds >> 24) as u8;
+            self.push_bytes(&encoding.to_le_bytes()[..]);
+            self.push_byte(high_six);
+        } else {
+            self.push_bytes(&encoding.to_le_bytes()[..encoded_body_length]);
+        }
+
+        Ok(())
+    }
+
+    fn write_long_form_timestamp(mut self, value: &Timestamp) -> IonResult<()> {
+        use crate::TimestampPrecision::*;
+
+        // The number of bits dedicated to each time unit in a long-form timestamp
+        const NUM_YEAR_BITS: u32 = 14;
+        const NUM_MONTH_BITS: u32 = 4;
+        const NUM_DAY_BITS: u32 = 5;
+        const NUM_HOUR_BITS: u32 = 5;
+        const NUM_MINUTE_BITS: u32 = 6;
+        const NUM_SECOND_BITS: u32 = 6;
+        // The number of bits dedicated to representing the offset in a long-form timestamp
+        const NUM_OFFSET_BITS: u32 = 12;
+        // The bit offsets of each time unit within the encoding
+        const MONTH_BIT_OFFSET: u32 = NUM_YEAR_BITS;
+        const DAY_BIT_OFFSET: u32 = MONTH_BIT_OFFSET + NUM_MONTH_BITS;
+        const HOUR_BIT_OFFSET: u32 = DAY_BIT_OFFSET + NUM_DAY_BITS;
+        const MINUTE_BIT_OFFSET: u32 = HOUR_BIT_OFFSET + NUM_HOUR_BITS;
+        const OFFSET_BIT_OFFSET: u32 = MINUTE_BIT_OFFSET + NUM_MINUTE_BITS;
+        const SECOND_BIT_OFFSET: u32 = OFFSET_BIT_OFFSET + NUM_OFFSET_BITS;
+
+        let mut encoding: u64 = value.year() as u64;
+        encoding |= (value.month() as u64) << MONTH_BIT_OFFSET;
+        encoding |= (value.day() as u64) << DAY_BIT_OFFSET;
+        encoding |= (value.hour() as u64) << HOUR_BIT_OFFSET;
+        encoding |= (value.minute() as u64) << MINUTE_BIT_OFFSET;
+
+        const MIN_OFFSET: i32 = -24 * 60;
+        const UNKNOWN_OFFSET: u64 = (1 << NUM_OFFSET_BITS) - 1;
+
+        let offset_minutes = match value.offset() {
+            Some(minutes) => (minutes - MIN_OFFSET) as u64,
+            None => UNKNOWN_OFFSET,
+        };
+        encoding |= offset_minutes << OFFSET_BIT_OFFSET;
+        encoding |= (value.second() as u64) << SECOND_BIT_OFFSET;
+
+        let precision = value.precision();
+        let scale = value.fractional_seconds_scale().unwrap_or(0);
+
+        // The encoded bytelength of all components *except* subseconds.
+        let (mut encoded_length, num_bits_in_use) = match precision {
+            Year => (2, 14),
+            Month => (3, MONTH_BIT_OFFSET + NUM_MONTH_BITS),
+            Day => {
+                let bits_in_use = DAY_BIT_OFFSET + NUM_DAY_BITS;
+                // Set the "day-not-month" bit, just beyond the day bits.
```

Comment:
I'm not sure I understand why this is necessary. If there's "day" precision, then the day will already be non-zero. If anything, shouldn't we be zero-ing the day if it's month precision?

Reply:
> If anything, shouldn't we be zero-ing the day if it's month precision?

This is related to the question raised by https://github.com/amazon-ion/ion-docs/issues/295; if the writer MUST zero out unused bits, then the reader could use a `day` of `0` as an indication. However, if the reader MUST ignore unused bits, then we need a dedicated bit.

I've added your question to that issue and made the issue more general.